frontend/src/components/nodes/NodesMap.tsx

```diff
@@ -0,0 +1,791 @@
+        peerCount: "",
+        radius: 0,
+        fillKey: "",
+        agentName: "",
+        version: "",
+        build: "",
+        nodeId: "",
+        blockNr: 0,
+        lastSeen: 0,
+        location: "",
+      };
+
+      bubble.latitude = geoData[index].lat;
+      bubble.longitude = geoData[index].lon;
+      bubble.location = geoData[index].city;
+      bubble.name = element.accountId;
+      bubble.isValidator = element.isValidator;
+      bubble.peerCount = element.peerCount;
+      bubble.agentName = element.agentName;
+      bubble.version = element.agentVersion;
+      bubble.build = element.agentBuild;
+      bubble.nodeId = element.nodeId;
+      bubble.blockNr = element.lastHeight;
+      bubble.lastSeen = element.lastSeen;
+      bubble.radius = 4;
+      bubble.fillKey = element.isValidator
+        ? "validatorBubbleFill"
+        : "nonValidatorBubbleFill";
+
+      return bubble;
+    });
+
+    this.saveData(bubbles);
+  }
+
+  saveData(newNodes: IBubble[]) {
+    const oldNodes = this.state.nodesData;
+    const newAddedNodes: IBubble[] = newNodes.filter(
+      this.compareObjectsArrays(oldNodes)
+    );
+    const removedNodes: IBubble[] = oldNodes.filter(
+      this.compareObjectsArrays(newNodes)
+    );
+    const groupedNodes = this.getNodeClusters(newNodes);
+
+    this.setState({
+      nodesData: newNodes,
+      newNodes: newAddedNodes,
+      removedNodes: removedNodes,
+      nodeClusters: groupedNodes,
+    });
+  }
+
+  compareObjectsArrays(objectArray: IBubble[]) {
+    return (current: IBubble) => {
+      return (
+        objectArray.filter((other) => {
+          return other.nodeId == current.nodeId;
+        }).length == 0
+      );
+    };
+  }
+
+  getNodeClusters(nodes: IBubble[]) {
+    // Get clusters of nodes that share the same location.
+    const groupedNodes: any = lodash.groupBy(nodes, (item: IBubble) => {
+      return item.latitude;
+    });
+    // @ts-ignore
+    lodash.forEach(groupedNodes, (value, key) => {
+      // value is unused but necessary, without it function gets weird behaviour
+      groupedNodes[key] = lodash.groupBy(groupedNodes[key], (item) => {
+        return item.longitude;
+      });
+    });
+
+    const finalArray: any[] = [];
+    Object.keys(groupedNodes).forEach((element) => {
+      Object.keys(groupedNodes[element]).forEach((item) => {
+        if (groupedNodes[element][item].length > 1) {
+          finalArray.push(groupedNodes[element][item]);
+        }
+      });
+    });
+    return finalArray;
+  }
+
+  changeToValidators() {
+    this.setState(
+      {
+        nodesType: "validators",
+        newNodes: [],
+      },
+      () => this.fetchGeo()
+    );
+  }
+
+  changeToNonValidators() {
+    this.setState(
+      {
+        nodesType: "non-validators",
+        newNodes: [],
+      },
+      () => this.fetchGeo()
+    );
+  }
+
+  renderBubbleTooltip(data: IBubble) {
+    // Prettier ruined the entire indentation that i made for this
```

Comment:
so sorry for this. Do we have other options to do the tooltip, like using some existing libraries to help? this is a little hard to read and change.

Reply:
unfortunately the answer is no :( or not yet, i tried to use some jsx and convert it to string with npm jsx-to-string but didn't went well

Reply:
i have an idea tho, i'll try it tomorrow. now is 2:58AM

Reply:
Try using formatted strings: `<div>${data.name}</div>`

Reply:
https://github.com/near/near-explorer/pull/335/commits/d9ff1face1e64bea0fc58bf9fe2a586e904aa8de -- replaced 
many one liner formatted strings with single formatted string and looks a little better